src/mono/mono/metadata/appdomain.c

```diff
@@ -774,6 +774,12 @@ ves_icall_System_Reflection_Assembly_InternalLoad (MonoStringHandle name_handle,
 	MonoAssembly *requesting_assembly = mono_runtime_get_caller_from_stack_mark (stack_mark);
 	MonoAssemblyLoadContext *alc = (MonoAssemblyLoadContext *)load_Context;
 
+#if HOST_WASI
+	// On WASI, mono_assembly_get_alc isn't supported
```

Comment:
Is this a wasi limitation, or a general problem ?

Reply:
I've updated the comment here to clarify that it's not a permanent limitation but rather is something that remains to be done.

Reply:
> `mono_image_get_alc` fails with `Error: wasm trap: out of bounds memory access.`

That's not great.  Probably either `image` is null or it's junk.  Which means the assembly passed to `mono_assembly_get_alc` is null or junk.

So we're masking some underlying issue.   Does the trap happen early in startup or once the app is running?  can you build the driver with logging turned on (setenv `MONO_LOG_MASK=asm`, `MONO_LOG_LEVEL=debug`) - not sure if that will give any clues, but it might give an idea of when it's happening.

Reply:
It happens during startup when trying to load the application's entrypoint assembly. It doesn't happen for a basic console app but does for an ASP.NET Core one, so it's something triggered inside ASP.NET Core, perhaps as it's trying to do some reflection over the entrypoint assembly.

The issue is that `mono_runtime_get_caller_from_stack_mark` on line 774 above is returning `NULL`, and so the attempt to dereference `assm->image` inside `mono_assembly_get_alc` will fail. I suppose internally it's something about stack walking being incomplete.

I'll add this to a list of things needing to be tracked down for complete WASI support in the future.